cloud/log_syncer.go

```diff
@@ -0,0 +1,203 @@
+
+// DefaultLogBatchSize is the default batch size.
+const DefaultLogBatchSize = 256
+
+// DefaultLogIdleDuration is the maximum idle duration before a sync could happen.
+const DefaultLogIdleDuration = 1 * time.Second
+
+// NewLogSyncer creates a new log syncer.
+func NewLogSyncer(syncfn Syncer) *LogSyncer {
+	return NewLogSyncerWith(syncfn, DefaultLogMaxLineSize, DefaultLogBatchSize, DefaultLogIdleDuration)
+}
+
+// NewLogSyncerWith creates a new customizable syncer.
+func NewLogSyncerWith(
+	syncfn Syncer,
+	maxLineSize int,
+	batchSize int,
+	idleDuration time.Duration,
+) *LogSyncer {
+	if maxLineSize == 0 {
+		panic("max line size must be set")
+	}
+	l := &LogSyncer{
+		in:       make(chan *DeploymentLog, batchSize),
+		syncfn:   syncfn,
+		shutdown: make(chan struct{}),
+
+		maxLineSize:  maxLineSize,
+		batchSize:    batchSize,
+		idleDuration: idleDuration,
+	}
+	l.start()
+	return l
+}
+
+// NewBuffer creates a new synchronized buffer.
+func (s *LogSyncer) NewBuffer(channel LogChannel, out io.Writer) io.Writer {
+	r, w := io.Pipe()
+	s.fds = append(s.fds, w)
+	s.wg.Add(1)
+	go func() {
+		defer s.wg.Done()
+		line := int64(1)
+		scanner := bufio.NewScanner(r)
+		maxLineSize := s.maxLineSize
+		buf := make([]byte, maxLineSize)
+		scanner.Buffer(buf, maxLineSize)
+		// no internal allocation
+		scanner.Split(scanLines)
+
+		errs := errors.L()
+		for scanner.Scan() {
+			message := scanner.Text()
+			_, err := out.Write([]byte(message))
+			if err != nil {
+				errs.Append(errors.E(err, "writing to terminal"))
+				continue
+			}
+
+			t := time.Now().UTC()
+			s.in <- &DeploymentLog{
+				Channel:   channel,
+				Line:      line,
+				Message:   string(dropCRLN([]byte(message))),
+				Timestamp: &t,
+			}
+			line++
+		}
+		if err := scanner.Err(); err != nil {
+			errs.Append(errors.E(err, "scanning output lines"))
+		}
+
+		errs.Append(r.Close())
+		errs.Append(w.Close())
+		if err := errs.AsError(); err != nil {
+			log.Error().Err(err).Msg("synchroning command output")
+		}
+	}()
+	return w
+}
+
+// Wait waits for the processing of all log messages.
+// After calling this method, it's not safe to call any other method, as it
+// closes the internal channels and shutdown all goroutines.
+func (s *LogSyncer) Wait() {
+	for _, writerFD := range s.fds {
+		// only return an error when readerFD.CloseWithError(err) is called but
+		// but this is not the case.
+		_ = writerFD.Close()
+	}
+	s.wg.Wait()
+	close(s.in)
+	<-s.shutdown
+}
+
+func (s *LogSyncer) start() {
+	go func() {
+		s.lastEnqueued = time.Now()
+		for e := range s.in {
+			s.enqueue(e)
+		}
+		for len(s.pending) > 0 {
+			rest := min(s.batchSize, len(s.pending))
+			s.syncfn(s.pending[:rest])
+			s.pending = s.pending[rest:]
+		}
+		s.shutdown <- struct{}{}
+	}()
+}
+
+func (s *LogSyncer) enqueue(l *DeploymentLog) {
+	s.pending = append(s.pending, l)
+	if len(s.pending) >= s.batchSize ||
+		(len(s.pending) > 0 && time.Since(s.lastEnqueued) > s.idleDuration) {
```

Comment:
Do I understanding correctly that the buffered logs are only sent after either idle time or batch size are exceeded? If yes, that means at worst case there could be a ~256s delay if a command keeps adding output ~every second. Even without the worst case, for commands that keep producing output it would probably mean that syncing is delayed until they're done.
If we want to provide a "live" view of the output in the cloud UI, we could consider using a fixed interval instead.

Reply:
As a follow-up, here's an idea for an implementation that syncs in a fixed interval:
```
func (s *LogSyncer) start() {
	go func() {
		ticker := time.NewTicker(500 * time.Millisecond)

		defer ticker.Stop()
		defer func() { s.shutdown <- struct{}{} }()

		for range ticker.C {
			next, done := s.fetchNext()
			if len(next) > 0 {
				s.syncfn(next)
			}
			if done {
				return
			}
		}
	}()
}

func (s *LogSyncer) fetchNext() (next DeploymentLogs, done bool) {
	for {
		select {
		case m, ok := <-s.in:
			// Keep taking elements from the queue ...
			if ok {
				next = append(next, m)

				// ... until batch size is exceeded
				if len(next) >= s.batchSize {
					return next, false
				}
			} else {
				// ... or until the queue has been closed
				return next, true
			}

		default:
			// ... or until there are no more elements for now
			return next, false
		}
	}
}
```

Reply:
As aligned in other channels, it definitely makes sense to sync at regular intervals so the user have a nicer UI feedback to the command output, but we will improve this behavior in a separate PR in next release.

Thanks @snakster